File: apps/mobile/ios-native/AppServerWebSocketClient.swift
import Foundation

actor AppServerWebSocketClient {
    private let webSocketURL: URL
    private let headers: [String: String]
    private let onNotification: @Sendable (String, JSONDictionary) async -> Void
    private let onLog: @Sendable (String, Error?) -> Void

    private var nextRequestID = 2
    private var pendingRequests: [Int: CheckedContinuation<JSONDictionary, Error>] = [:]
    private var session: URLSession?
    private var socket: URLSessionWebSocketTask?
    private var readyTask: Task<Void, Error>?
    private var receiveLoopTask: Task<Void, Never>?

    init(
        webSocketURL: URL,
        headers: [String: String],
        onNotification: @escaping @Sendable (String, JSONDictionary) async -> Void,
        onLog: @escaping @Sendable (String, Error?) -> Void
    ) {
        self.webSocketURL = webSocketURL
        self.headers = headers
        self.onNotification = onNotification
        self.onLog = onLog
    }

    func request(method: String, params: JSONDictionary) async throws -> JSONDictionary {
        try await ensureConnected()
        let requestID = nextRequestID
        nextRequestID += 1
        return try await withCheckedThrowingContinuation { continuation in
            pendingRequests[requestID] = continuation
            Task {
                do {
                    try await sendPayload([
                        "jsonrpc": "2.0",
                        "id": requestID,
                        "method": method,
                        "params": params,
                    ])
                } catch {
                    failRequest(id: requestID, error: error)
                }
            }
        }
    }

    func close() async {
        failAll(with: NativeHostBridgeError.webSocketClosed("App server websocket closed."))
        receiveLoopTask?.cancel()
        socket?.cancel(with: .normalClosure, reason: nil)
        session?.invalidateAndCancel()
        socket = nil
        session = nil
        readyTask = nil
    }

    private func ensureConnected() async throws {
        if let readyTask {
            try await readyTask.value
            return
        }
        let task = Task<Void, Error> {
            try await connect()
        }
        readyTask = task
        do {
            try await task.value
        } catch {
            readyTask = nil
            throw error
        }
    }

    private func connect() async throws {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 15
        configuration.timeoutIntervalForResource = 15
        let session = URLSession(configuration: configuration)
        var request = URLRequest(url: webSocketURL)
        headers.forEach { request.setValue($1, forHTTPHeaderField: $0) }
        let socket = session.webSocketTask(with: request)

        self.session = session
        self.socket = socket
        socket.resume()
        receiveLoopTask = Task { [weak self] in
            guard let self else { return }
            await self.receiveLoop(using: socket)
        }

        _ = try await sendRequestDuringConnection(
            id: 1,
            method: "initialize",
            params: [
                "clientInfo": [
                    "name": "codex_mobile_host",
                    "title": "Codex Mobile Host",
                    "version": "0.1.0",
                ],
                "capabilities": [
                    "experimentalApi": true,
                ],
            ]
        )

        try await sendPayload([
            "jsonrpc": "2.0",
            "method": "initialized",
            "params": [:],
        ])
    }

    private func sendRequestDuringConnection(id: Int, method: String, params: JSONDictionary) async throws -> JSONDictionary {
        try await withCheckedThrowingContinuation { continuation in
            pendingRequests[id] = continuation
            Task {
                do {
                    try await sendPayload([
                        "jsonrpc": "2.0",
                        "id": id,
                        "method": method,
                        "params": params,
                    ])
                } catch {
                    failRequest(id: id, error: error)
                }
            }
        }
    }

    private func sendPayload(_ payload: JSONDictionary) async throws {
        guard let socket else {
            throw NativeHostBridgeError.webSocketClosed("App server websocket is not connected.")
        }
        try await socket.send(.string(jsonEncodedString(payload)))
    }

    private func receiveLoop(using socket: URLSessionWebSocketTask) async {
        do {
            while !Task.isCancelled {
                let message = try await socket.receive()
                let text: String
                switch message {
                case let .string(value):
                    text = value
                case let .data(data):
                    text = String(data: data, encoding: .utf8) ?? ""
                @unknown default:
                    continue
                }
                await handleMessage(text)
            }
        } catch {
            onLog("app server websocket failure", error)
            failAll(with: error)
            readyTask = nil
            self.socket = nil
        }
    }

    private func handleMessage(_ text: String) async {
        guard let payload = parseJSONValue(from: text) as? JSONDictionary else {
            return
        }
        if let id = payload["id"] as? Int {
            handleResponse(id: id, payload: payload)
            return
        }

        guard let method = payload["method"] as? String,
              !method.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return
        }
        let params = payload["params"] as? JSONDictionary ?? [:]
        await onNotification(method, deepCopyJSONObject(params))
    }

    private func handleResponse(id: Int, payload: JSONDictionary) {
        guard let continuation = pendingRequests.removeValue(forKey: id) else {
            return
        }
        if let errorObject = payload["error"] as? JSONDictionary {
            let trimmedMessage = (errorObject["message"] as? String)?
                .trimmingCharacters(in: .whitespacesAndNewlines)
            let message = trimmedMessage?.isEmpty == false ? trimmedMessage! : "App server request failed."
            continuation.resume(throwing: NativeHostBridgeError.requestFailed(message))
            return
        }
        let result = payload["result"] as? JSONDictionary ?? [:]
        continuation.resume(returning: deepCopyJSONObject(result))
    }

    private func failRequest(id: Int, error: Error) {
        guard let continuation = pendingRequests.removeValue(forKey: id) else {
            return
        }
        continuation.resume(throwing: error)
    }

    private func failAll(with error: Error) {
        let requests = pendingRequests
        pendingRequests.removeAll()
        requests.values.forEach { continuation in
            continuation.resume(throwing: error)
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/boomyao/codexmobile/nativehost/AppServerWebSocketClient.kt
package com.boomyao.codexmobile.nativehost

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

class AppServerWebSocketClient(
    private val webSocketUrl: String,
    private val headers: Map<String, String>,
    private val onNotification: (String, JSONObject) -> Unit,
    private val onLog: (String, Throwable?) -> Unit,
    private val onDisconnected: (Throwable?) -> Unit,
) : WebSocketListener() {
    private val httpClient =
        OkHttpClient.Builder()
            .pingInterval(15, TimeUnit.SECONDS)
            .readTimeout(0, TimeUnit.MILLISECONDS)
            .build()
    private val nextRequestId = AtomicInteger(2)
    private val pendingRequests = ConcurrentHashMap<Int, CompletableFuture<JSONObject>>()
    private val readyFuture = CompletableFuture<Unit>()

    @Volatile
    private var socket: WebSocket? = null

    @Volatile
    private var connectStarted = false

    @Volatile
    private var closeRequested = false

    @Volatile
    private var disconnectNotified = false

    fun matches(url: String, extraHeaders: Map<String, String>): Boolean {
        return webSocketUrl == url && headers == extraHeaders
    }

    fun request(method: String, params: JSONObject): JSONObject {
        ensureConnected()
        readyFuture.get(15, TimeUnit.SECONDS)

        val requestId = nextRequestId.getAndIncrement()
        val resultFuture = CompletableFuture<JSONObject>()
        pendingRequests[requestId] = resultFuture

        val payload =
            JSONObject()
                .put("jsonrpc", "2.0")
                .put("id", requestId)
                .put("method", method)
                .put("params", JSONObject(params.toString()))

        val currentSocket = socket ?: throw IOException("App server websocket is not connected.")
        if (!currentSocket.send(payload.toString())) {
            pendingRequests.remove(requestId)
            val error = IOException("Failed to write app server websocket request.")
            notifyDisconnected(error)
            throw error
        }

        return resultFuture.get(10, TimeUnit.MINUTES)
    }

    fun close() {
        closeRequested = true
        failAll(IOException("App server websocket closed."))
        socket?.close(1000, "closing")
        socket = null
        httpClient.dispatcher.executorService.shutdown()
        httpClient.connectionPool.evictAll()
    }

    override fun onOpen(webSocket: WebSocket, response: Response) {
        socket = webSocket
        val initialize =
            JSONObject()
                .put("jsonrpc", "2.0")
                .put("id", 1)
                .put("method", "initialize")
                .put(
                    "params",
                    JSONObject()
                        .put(
                            "clientInfo",
                            JSONObject()
                                .put("name", "codex_mobile_host")
                                .put("title", "Codex Mobile Host")
                                .put("version", "0.1.0"),
                        )
                        .put("capabilities", JSONObject().put("experimentalApi", true)),
                )
        webSocket.send(initialize.toString())
    }

    override fun onMessage(webSocket: WebSocket, text: String) {
        val payload = JSONObject(text)
        if (payload.has("id")) {
            handleResponse(webSocket, payload)
            return
        }

        val method = payload.optString("method").trim()
        if (method.isEmpty()) {
            return
        }
        val params = payload.optJSONObject("params") ?: JSONObject()
        onNotification(method, JSONObject(params.toString()))
    }

    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        val error = IOException("App server websocket closed: $code $reason")
        failAll(error)
        socket = null
        notifyDisconnected(error)
    }

    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        onLog("app server websocket failure", t)
        failAll(t)
        socket = null
        notifyDisconnected(t)
    }

    private fun ensureConnected() {
        if (socket != null || connectStarted) {
            return
        }

        synchronized(this) {
            if (socket != null || connectStarted) {
                return
            }
            connectStarted = true
            val requestBuilder = Request.Builder().url(webSocketUrl)
            headers.forEach { (key, value) ->
                requestBuilder.addHeader(key, value)
            }
            httpClient.newWebSocket(requestBuilder.build(), this)
        }
    }

    private fun handleResponse(webSocket: WebSocket, payload: JSONObject) {
        val id = payload.optInt("id", -1)
        if (id == 1) {
            if (payload.has("error")) {
                val message =
                    payload.optJSONObject("error")
                        ?.optString("message")
                        ?.trim()
                        .orEmpty()
                        .ifEmpty { "Failed to initialize app server websocket." }
                readyFuture.completeExceptionally(IOException(message))
                return
            }
            webSocket.send(
                JSONObject()
                    .put("jsonrpc", "2.0")
                    .put("method", "initialized")
                    .put("params", JSONObject())
                    .toString(),
            )
            readyFuture.complete(Unit)
            return
        }

        val future = pendingRequests.remove(id) ?: return
        if (payload.has("error")) {
            val message =
                payload.optJSONObject("error")
                    ?.optString("message")
                    ?.trim()
                    .orEmpty()
                    .ifEmpty { "App server request failed." }
            future.completeExceptionally(IOException(message))
            return
        }

        val result = payload.optJSONObject("result") ?: JSONObject()
        future.complete(JSONObject(result.toString()))
    }

    private fun failAll(error: Throwable) {
        if (!readyFuture.isDone) {
            readyFuture.completeExceptionally(error)
        }
        val iterator = pendingRequests.entries.iterator()
        while (iterator.hasNext()) {
            val entry = iterator.next()
            iterator.remove()
            entry.value.completeExceptionally(error)
        }
    }

    private fun notifyDisconnected(error: Throwable?) {
        if (closeRequested || disconnectNotified) {
            return
        }

        synchronized(this) {
            if (closeRequested || disconnectNotified) {
                return
            }
            disconnectNotified = true
        }
        onDisconnected(error)
    }
}
